simtools/corsika/corsika_output.py

```diff
@@ -0,0 +1,1301 @@
+import functools
+import logging
+import operator
+import time
+from pathlib import Path
+
+import boost_histogram as bh
+import numpy as np
+from astropy import units as u
+from astropy.units import cds
+from corsikaio.subblocks import event_header, get_units_from_fields, run_header
+from eventio import IACTFile
+
+from simtools import io_handler
+from simtools.util.general import (
+    collect_data_from_yaml_or_dict,
+    convert_2D_to_radial_distr,
+    rotate,
+    save_dict_to_file,
+)
+
+
+class HistogramNotCreated(Exception):
+    """Exception for histogram not created."""
+
+
+class CorsikaOutput:
+    """CorsikaOutput extracts the Cherenkov photons information from a CORSIKA output file
+    (IACT file) using pyeventio.
+
+    Parameters
+    ----------
+    input_file: str or Path
+        Input file (IACT file) provided by the CORSIKA simulation.
+    label: str
+        Instance label.
+
+    Raises
+    ------
+    FileNotFoundError:
+        if the input file given does not exist.
+    """
+
+    def __init__(self, input_file, label=None):
+        self.label = label
+        self._logger = logging.getLogger(__name__)
+        self._logger.debug("Init CorsikaOutput")
+        self.input_file = input_file
+
+        self.input_file = Path(self.input_file)
+        if not self.input_file.exists():
+            msg = f"file {self.input_file} does not exist."
+            self._logger.error(msg)
+            raise FileNotFoundError
+
+        self.io_handler = io_handler.IOHandler()
+
+        self._initialize_attributes()
+        self.read_event_information()
+        self._initialize_header()
+
+    def _initialize_attributes(self):
+        """
+        Initializes the class attributes.
+        """
+        self._telescope_indices = None
+        self._telescope_positions = None
+        self.num_events = None
+        self.num_of_hist = None
+        self.num_telescopes = None
+        self._num_photons_per_event_per_telescope = None
+        self._num_photons_per_event = None
+        self._num_photons_per_telescope = None
+        self._event_azimuth_angles = None
+        self._event_zenith_angles = None
+        self._hist_config = None
+        self._total_num_photons = None
+        self._magnetic_field_x = None
+        self._magnetic_field_z = None
+        self._event_total_energies = None
+        self._event_first_interaction_heights = None
+        self._version = None
+        self.event_information = None
+        self._individual_telescopes = None
+        self._allowed_histograms = {"hist_position", "hist_direction", "hist_time_altitude"}
+        self._allowed_1D_labels = {"wavelength", "time", "altitude"}
+        self._allowed_2D_labels = {"counts", "density", "direction", "time_altitude"}
+        self._header = None
+
+    @property
+    def version(self):
+        """
+        Get the version of the Corsika output file.
```

Comment:
Here again: could call it `CORSIKA IACT file` or `CORSIKA file`, as this output is the input for this class.